crates/composer/src/main.rs

```diff
@@ -0,0 +1,37 @@
+use composer::api::Event;
+use composer::api::DEFAULT_SERVER_ADDRESS;
+use eyre::Result;
+use rodio::{source::Source, Decoder, OutputStream};
+use std::fs::File;
+use std::io::BufReader;
+use std::net::UdpSocket;
+
+fn main() -> Result<()> {
+    color_eyre::install()?;
+
+    let socket = UdpSocket::bind(DEFAULT_SERVER_ADDRESS)?;
+    println!("Listening on {}", socket.local_addr()?);
+
+    let (_stream, stream_handle) = OutputStream::try_default()?;
+
+    loop {
+        if let Err(err) = {
+            // Should be more than enough.
+            let mut buf = [0; 1000];
+            let (number_of_bytes, _) = socket.recv_from(&mut buf)?;
+
+            let event: Event = bincode::deserialize(&buf)?;
+            println!("Received an event ({number_of_bytes} bytes): {:?}", event);
+
+            // FIXME: do the decoding and file reading outside the loop
+            let file = BufReader::new(File::open("src/sound_samples/click.wav")?);
+            let source = Decoder::new(file)?;
+            stream_handle.play_raw(source.convert_samples())?;
+
+            Ok::<(), eyre::Error>(())
+        } {
+            eprintln!("Could not process data-gram: {:?}", err);
+            continue;
+        }
+    }
```

Comment:
I may be completely misreading this (in which case I blame pre-coffee Friday brain) but I think this control flow is not right. The `continue` line should be impossible to hit under the current circumstances.

There are many fallible functions inside the `if let` block, and they're all postfixed with the `?` try operator. I feel like the intention was that any of these errors would bubble up to the `if let` and take control flow to the `continue` block. However, that's not how it would play out; any line erroring inside the `if let` block would instead exit the `main` function and terminate the program.

The only way I can see the `continue` block being hit is if the previous block evaluated to an `Err`, which is impossible as line `31` is an inconditional `Ok`.

A way to achieve what (I think) this function wants to do would be to take the block into a separate function, and use its return value. Then it would be OK to use all these `try` operators as-is.

Reply:
@PabloMansanet great catch! I completely missed that.

I think this is written as if `try { ... }` expressions from https://github.com/rust-lang/rust/issues/31436 where stable - but they are not (and this doesn't contain the critical `try` keyword).

Reply:
Good catch!

> I think this is written as if try { ... } expressions from https://github.com/rust-lang/rust/issues/31436 where stable

This semantic was exactly my intention. .)

I refactored this into a separate function in 5a375640f9e1e471bf9fc2fe618e8672dfdc148c